File: extra_learning/design_patterns/03_builder.py
"""Builder with fluent interface."""
from __future__ import annotations
from dataclasses import dataclass, field


@dataclass
class Query:
    table: str
    columns: list[str] = field(default_factory=list)
    where: str | None = None
    limit: int | None = None


class QueryBuilder:
    def __init__(self, table: str) -> None:
        self._q = Query(table=table)

    def select(self, *cols: str) -> "QueryBuilder":
        self._q.columns = list(cols); return self

    def where(self, expr: str) -> "QueryBuilder":
        self._q.where = expr; return self

    def limit(self, n: int) -> "QueryBuilder":
        self._q.limit = n; return self

    def build(self) -> Query:
        return self._q


if __name__ == "__main__":
    q = QueryBuilder("orders").select("id", "amount").where("amount > 100").limit(10).build()
    print(q)

File: go_for_data_engineering/05_design_patterns/03_builder.go
// Builder for complex objects with many optional fields.
// In Go, the functional-options pattern is usually preferred over a Builder struct.
package main

import "fmt"

type Query struct {
	Table   string
	Columns []string
	Where   string
	Limit   int
}

type QueryBuilder struct{ q Query }

func New(table string) *QueryBuilder                       { return &QueryBuilder{q: Query{Table: table}} }
func (b *QueryBuilder) Select(c ...string) *QueryBuilder   { b.q.Columns = c; return b }
func (b *QueryBuilder) WhereClause(w string) *QueryBuilder { b.q.Where = w; return b }
func (b *QueryBuilder) Limit(n int) *QueryBuilder          { b.q.Limit = n; return b }
func (b *QueryBuilder) Build() Query                       { return b.q }

func main() {
	q := New("orders").Select("id", "amount").WhereClause("amount > 100").Limit(10).Build()
	fmt.Printf("%+v\n", q)
}
